design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,582 @@
+administration console, and therefore by extension to how Keycloak handles authentication
+on a logical level. The design may also specify functions, data structures and screen
+mockups as the design progresses.
+
+The design of the step up authentication ([KEYCLOAK-847](https://issues.jboss.org/browse/KEYCLOAK-847)) process is also defined in this proposal as it is linked to the multi-factor
+authentication (MFA) and the mechanisms that are defined in the one will have an impact
+on the other.
+
+The design also has the following constraints:
+* The design must be simple to use, but can be customised for more complex use cases. This is true for both the administrator when setting up the realm and the user when logging in with Keycloak.
+* In addition to two factor authentication, the design must take into account the use with passwordless authentication (i.e. a single factor authentication that is not a password).
+* The system must be easily extensible for groups that wish to add their own authenticators within Keycloak.
+
+For [KEYCLOAK-9693](https://issues.jboss.org/browse/KEYCLOAK-9693) this means changes
+to the authentication logic, changes to the authentication part of the admin console,
+changes to the authentication screens that the user sees during login, changes to the
+REST API, and changes to the database.
+
+For [KEYCLOAK-9694](https://issues.jboss.org/browse/KEYCLOAK-9694) this means
+changes to the *users > credentials* part of the administration console and to
+the REST API.
+
+For [KEYCLOAK-847](https://issues.jboss.org/browse/KEYCLOAK-847) this means changes
+to the authentication logic, changes to the authentication part of the admin console,
+changes to the client part of the admin console, extending Keycloak's comprehension
+of the OIDC and SAML protocols, and some small changes to the REST API.
+
+What this design proposal does _not_ cover is the specific handling of the WebAuthn
+and modifications to the account console, as other design proposals and JIRAs already
+specify those modifications.
+
+## Design proposal for [KEYCLOAK-9693](https://issues.jboss.org/browse/KEYCLOAK-9693)
+
+In this section we discuss modifications necessary to allow an admin to manage two factor
+authentication, and the design of the admin GUI for this task. We will also be using the following concepts:
+* **Authentication factor**: A new type of execution in the authentication flow, to which only authentication executions can be assigned. This is designed to allow N factors to be added to an authentication flow.
+* **Authentication execution**: An execution that can only be set within an authentication factor. This is the actual code that is run to authenticate a user. Examples of authentication executions in Keycloak would be the classes OTPFormAuthenticator, UsernamePasswordForm or ValidateOTP.
+* **Authentication category**: A logical grouping between authentication executions and credentials. This design is to allow for multi-tokens.
```

Comment:
Do you mean "associating multiple OTP devices with an account" by "multi-token"?

Reply:
Yes, but not necessarily linked to multiple OTP devices. It could also be for multiple fido2 devices, for example.

Reply:
I get it